diff -urNp stunnel-5.18/configure.ac stunnel-5.18-patch/configure.ac
--- stunnel-5.18/configure.ac	2015-06-01 10:25:32.000000000 -0400
+++ stunnel-5.18-patch/configure.ac	2015-06-22 11:34:50.244471487 -0400
@@ -274,7 +274,7 @@ AC_ARG_ENABLE(fips,
 [  --disable-fips          disable OpenSSL FIPS support],
     [
         case "$enableval" in
-            yes) AC_MSG_RESULT([no])
+            yes) AC_MSG_RESULT([yes])
                  use_fips="yes"
                  AC_DEFINE([USE_FIPS], [1],
                     [Define to 1 to enable OpenSSL FIPS support])
