diff -up stunnel-5.42/src/options.c.system-ciphers stunnel-5.42/src/options.c
--- stunnel-5.42/src/options.c.system-ciphers	2017-08-22 14:24:31.169102148 +0200
+++ stunnel-5.42/src/options.c	2017-08-22 14:25:11.104019405 +0200
@@ -224,7 +224,7 @@ static char *option_not_found=
     "Specified option name is not valid here";
 
 static char *stunnel_cipher_list=
-    "HIGH:!DH:!aNULL:!SSLv2";
+    "PROFILE=SYSTEM";
 
 /**************************************** parse commandline parameters */
 
