diff -up stunnel-5.30/doc/stunnel.html.in.authpriv stunnel-5.30/doc/stunnel.html.in
--- stunnel-5.30/doc/stunnel.html.in.authpriv	2015-12-05 22:53:08.000000000 +0100
+++ stunnel-5.30/doc/stunnel.html.in	2016-02-03 16:39:02.259239765 +0100
@@ -246,7 +246,7 @@
 
 <p>Level is one of the syslog level names or numbers emerg (0), alert (1), crit (2), err (3), warning (4), notice (5), info (6), or debug (7). All logs for the specified level and all levels numerically less than it will be shown. Use <i>debug = debug</i> or <i>debug = 7</i> for greatest debugging output. The default is notice (5).</p>
 
-<p>The syslog facility &#39;daemon&#39; will be used unless a facility name is supplied. (Facilities are not supported on Win32.)</p>
+<p>The syslog facility &#39;authpriv&#39; will be used unless a facility name is supplied. (Facilities are not supported on Win32.)</p>
 
 <p>Case is ignored for both facilities and levels.</p>
 
diff -up stunnel-5.30/doc/stunnel.pod.in.authpriv stunnel-5.30/doc/stunnel.pod.in
--- stunnel-5.30/doc/stunnel.pod.in.authpriv	2015-12-05 22:52:36.000000000 +0100
+++ stunnel-5.30/doc/stunnel.pod.in	2016-02-03 16:39:20.123338658 +0100
@@ -195,7 +195,7 @@ info (6), or debug (7).  All logs for th
 all levels numerically less than it will be shown.  Use I<debug = debug> or
 I<debug = 7> for greatest debugging output.  The default is notice (5).
 
-The syslog facility 'daemon' will be used unless a facility name is supplied.
+The syslog facility 'authpriv' will be used unless a facility name is supplied.
 (Facilities are not supported on Win32.)
 
 Case is ignored for both facilities and levels.
diff -up stunnel-5.30/doc/stunnel.8.in.authpriv stunnel-5.30/doc/stunnel.8.in
--- stunnel-5.30/doc/stunnel.8.in.authpriv	2015-12-05 22:53:08.000000000 +0100
+++ stunnel-5.30/doc/stunnel.8.in	2016-02-03 16:38:31.199067820 +0100
@@ -207,7 +207,7 @@ info (6), or debug (7).  All logs for th
 all levels numerically less than it will be shown.  Use \fIdebug = debug\fR or
 \&\fIdebug = 7\fR for greatest debugging output.  The default is notice (5).
 .Sp
-The syslog facility 'daemon' will be used unless a facility name is supplied.
+The syslog facility 'authpriv' will be used unless a facility name is supplied.
 (Facilities are not supported on Win32.)
 .Sp
 Case is ignored for both facilities and levels.
--- stunnel-5.40/src/options.c	2017-01-28 08:47:51.000000000 +0000
+++ stunnel-5.40/src/options.c.patch	2017-01-30 15:40:36.859007125 +0000
@@ -1554,8 +1554,12 @@
     case CMD_BEGIN:
         section->log_level=LOG_NOTICE;
 #if !defined (USE_WIN32) && !defined (__vms)
+#if defined(LOG_AUTHPRIV)
+	new_global_options.log_facility=LOG_AUTHPRIV;
+#else
         new_global_options.log_facility=LOG_DAEMON;
 #endif
+#endif
         break;
     case CMD_EXEC:
         if(strcasecmp(opt, "debug"))
