diff -up stunnel-5.46/src/options.c.system-ciphers stunnel-5.46/src/options.c
--- stunnel-5.46/src/options.c.system-ciphers	2018-05-29 08:58:03.601089886 +0200
+++ stunnel-5.46/src/options.c	2018-05-29 08:59:00.880244728 +0200
@@ -252,7 +252,7 @@ static char *option_not_found=
     "Specified option name is not valid here";
 
 static char *stunnel_cipher_list=
-    "HIGH:!aNULL:!SSLv2:!DH:!kDHEPSK";
+    "PROFILE=SYSTEM";
 
 /**************************************** parse commandline parameters */
 
